### PR TITLE
FEATURE: Only output existing images and allow upscaling

### LINKED ISSUE #33: Don't render srcset paths that are bigger than the original image
When the editor uses a small image and the fusion code has bigger sizes defined in the srcset property they are added to the srcset html attribute.
In our project we had strange effects because the browser was expecting an image with a specific width (as defined in srcset) but got an image with a different width than specified. The image in our case was not even upscaled. But even if it were it would make no sense to upscale an image.

This patch fixed the problem for us:

[kaleidoscope-remove-upscaled-srcsets.patch.txt](https://github.com/sitegeist/Sitegeist.Kaleidoscope/files/5942269/kaleidoscope-remove-upscaled-srcsets.patch.txt)

### PR DESCRIPTION
**Pull request by manuelmeister**
This is a draft pr to limit srcset output to only existing images. 

I'm not sure I feel comfortable merging it like this. Maybe we first need a PR to add tests, so we can ensure that everything  still works correctly.

Fixes #33

### CODE REVIEW

**Comment by mficzel** [comment]
I like the idea, more precisely i think the biggest generated image should be the largest possible image with the required aspect ratio if there is an aspect ratio defined, otherwise the original one.

**Comment by manuelmeister** [comment]
That's exactly what I've done.

I've also increased the minimum php version to 7.4 as it [did not support an extended return type](https://github.com/sitegeist/Sitegeist.Kaleidoscope/actions/runs/6653945993/job/18080982197#step:5:28).

**Comment by mficzel** [comment]
@manuelmeister thanks a lot for the pr. Need to find some time and mindset for a proper review as this is non trivial.

**Comment by manuelmeister** [comment]
@mficzel Can you have a look again?

**Comment by mficzel** [comment]
I do not like the allowUpscaling argument that is passed around all over the place and adds quite a bit of complexity while i think that upscaling general is not a good idea. Also this pr does some unrelated things that look valuable but should not be mixed in this one.

How about this approach:
1. Add a method. `supportsUpscaling(): bool` to the scalableImageSourceInterface
2. Asset Image source should return `false` while DummyImageSource should return `true` 
3. Skip srcset items that would yield a bigger image if the CurrentImageSource does not supportsUpscaling 

That way we have strong defaults and anyone that has a need for upscaling images can still create a custom AssetImageSource

**Comment by mficzel** [comment]
@manuelmeister i talked with @bweinzierl at the contribution day and he is willing to help here aswell. 

BTW: A thing that is important to me is that the pr only does what is mentioned. The cleanups you added are nice but make reviewing harder. It would be great to split those topics into separate prs.

**Comment by manuelmeister** [comment]
@mficzel I agree with your comments. I'm not sure how to continue with the preserve aspect ratio.

**Comment by manuelmeister** [comment]
@mficzel I removed the aspect ratio part from this PR and opened a new one:  #77

File: Classes/Domain/AbstractImageSource.php
Comment (mficzel) [inline Classes/Domain/AbstractImageSource.php:1]: Theroretically breaky if one implemented the scalableImageSourceInterface without using the AbstractScalableImageSource as a base ... i do not think this is a real case

**Review by mficzel — CHANGES_REQUESTED** [review]
I like this now. However do we really have to add the "supportUpscaling" flag? I see no real user case for that and would leave this out and always assume allowUpscalinj = false. Just to make the tests for us a little bit easier.

File: Classes/Domain/AbstractImageSource.php
Comment (manuelmeister) [reply Classes/Domain/AbstractImageSource.php:1]: I don't understand

Comment (mficzel) [reply Classes/Domain/AbstractImageSource.php:1]: Not that important. I meant I consider this theoretical and not a real problem. So this is fine.

**Comment by manuelmeister** [comment]
For a project we needed the image upscaling, to ensure the images are the same.

**Review by mficzel — COMMENTED** [review]
I generally approve. This makes sense and i am fine with the structure. Will do some extra testing before merging to be extra sure this is as non breaky as possible,

Thanks for the pr.

**Review by mficzel — APPROVED** [review]

**Pull request merged**